File: scrapers/money_control.py
import requests
from bs4 import BeautifulSoup
from typing import List
from datetime import datetime
from scrapers.article import Article
import os


def article_links(page_start: int, page_end: int) -> List[str]:
    article_links = []
    for i in range(page_start, page_end):
        base_url = "https://www.moneycontrol.com/news/tags/companies/news"
        url = f"{base_url}/page-{i}/"
        req = requests.request(url=url, method="GET")
        soup = BeautifulSoup(req.text, "lxml")
        articles_ul = soup.find("ul", id="cagetory")
        article_lis = articles_ul.find_all("li", class_="clearfix")
        for article_li in article_lis:
            ele = article_li.find("h2").find("a")["href"]
            article_links.append(ele)
        print(f"page {i} completed...")
    return article_links


def scrape_article(article_url: str) -> Article:
    print(f"scraping {article_url}...")
    req = requests.request(url=article_url, method="GET")
    soup = BeautifulSoup(req.text, "lxml")

    article = soup.find("div", class_="page_left_wrapper")

    title = article.find("h1").text
    desc = article.find("h2").text

    time_div = article.find("div", class_="article_schedule")
    date = time_div.find("span").text
    time = time_div.text.split("/")[-1].strip()
    datetime_obj = datetime.strptime(f"{date} {time}", "%B %d, %Y %H:%M IST")

    content_div = article.find("div", id="contentdata")
    if not content_div:
        print(f"article {article_url} contains no content, skipping...")
        return None

    ps = content_div.find_all("p", class_="")
    content = []
    for p in ps:
        content.append(p.text.strip())

    return Article(article_url, title.strip(), desc.strip(), " ".join(content), datetime_obj)


def scrape(page_start: int, page_end: int) -> List[Article]:
    articles = []
    print("scraping article list...")
    links = article_links(page_start, page_end)
    print("scraping articles...")
    for link in links:
        article = scrape_article(link)
        if article:
            articles.append(article)
    return articles


def scrape_and_save(page_start: int, page_end: int, dir_name: str):
    articles = scrape(page_start, page_end)
    os.mkdir(dir_name)
    print("dumping pkl files...")
    for article in articles:
        article.dump(f"{dir_name}/{article.time}-{article.title}")

File: scraper/money-control.go
package scraper

import (
	"context"
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"github.com/gocolly/colly"
	"github.com/papaya147/parallelize"
	"log/slog"
	"strings"
	"time"
)

type MoneyControlScraper struct {
	logger  *slog.Logger
	baseUrl string
}

func NewMoneyControlScraper(logger *slog.Logger) *MoneyControlScraper {
	return &MoneyControlScraper{
		logger:  logger,
		baseUrl: "https://www.moneycontrol.com/news/tags/companies/news",
	}
}

func (mcs *MoneyControlScraper) articleLinks(start, end int) []string {
	var articleLinks []string
	c := colly.NewCollector()

	c.OnHTML("ul#cagetory li.clearfix h2 a", func(e *colly.HTMLElement) {
		link := e.Attr("href")
		articleLinks = append(articleLinks, link)
	})

	c.OnError(func(r *colly.Response, err error) {
		mcs.logger.Error(err.Error())
	})

	for i := start; i <= end; i++ {
		url := fmt.Sprintf("%s/page-%d/", mcs.baseUrl, i)
		mcs.logger.Info(fmt.Sprintf("visiting page %d", i))

		err := c.Visit(url)
		if err != nil {
			mcs.logger.Error(fmt.Sprintf("error visiting page %d: %s", i, err))
		}

		c.Wait()
		mcs.logger.Info(fmt.Sprintf("page %d completed", i))
	}

	return articleLinks
}

func (mcs *MoneyControlScraper) Scrape(start, end int) Articles {
	links := mcs.articleLinks(start, end)
	var articles []*Article

	batchSize := 5

	for i := 0; i < len(links); i += batchSize {
		end = i + batchSize
		if end > len(links) {
			end = len(links)
		}

		batchLinks := links[i:end]

		g := parallelize.NewGroup()
		channels := make([]parallelize.WithOutputWithArgsChannels[*Article], len(batchLinks))
		for j, link := range batchLinks {
			ch := parallelize.AddWithOutputWithArgs(g, mcs.scrape, context.Background(), link)
			channels[j] = ch
		}

		g.Execute()

		for _, ch := range channels {
			article, err := ch.Read()
			if err != nil {
				mcs.logger.Error(err.Error())
			}

			if article != nil {
				articles = append(articles, article)
			}
		}
	}

	return articles
}

func (mcs *MoneyControlScraper) scrape(_ context.Context, link string) (*Article, error) {
	mcs.logger.Info(fmt.Sprintf("scraping link %s", link))
	c := colly.NewCollector()

	var title, desc, content string
	var timestamp time.Time
	var err error

	c.OnHTML("div.page_left_wrapper", func(e *colly.HTMLElement) {
		title = strings.TrimSpace(e.ChildText("h1"))
		desc = strings.TrimSpace(e.ChildText("h2"))

		timeDiv := e.DOM.Find("div.article_schedule")
		date := timeDiv.Find("span").Text()
		timeText := strings.TrimSpace(strings.Split(timeDiv.Text(), "/")[1])
		timestamp, err = time.Parse("January 02, 2006 15:04 MST", fmt.Sprintf("%s %s", date, timeText))
		if err != nil {
			mcs.logger.Error(err.Error())
		}

		contentDiv := e.DOM.Find("div#contentdata")
		if contentDiv.Length() == 0 {
			mcs.logger.Info(fmt.Sprintf("article %s contains no content, skipping", link))
			return
		}

		contentDiv.Find("p").Each(func(i int, selection *goquery.Selection) {
			attr, _ := selection.Attr("class")
			if attr == "" {
				content += strings.TrimSpace(selection.Text()) + " "
			}
		})
	})

	err = c.Visit(link)
	if err != nil {
		return nil, err
	}

	c.Wait()

	if content == "" {
		return nil, nil
	}

	return NewArticle(link, title, desc, content, timestamp), nil
}
